cirq/sim/wave_function.py

```diff
@@ -345,36 +357,137 @@ def to_valid_state_vector(
+                    'ambiguous whether or not the given `state_like` is a '
+                    'state vector or a list of computational basis values for '
+                    'the qudits. In this situation you are required to pass '
+                    'in a state vector that is a numpy array with a complex '
+                    'dtype.')
+
+        if state_like.shape == (prod,) or state_like.shape == qid_shape:
+            return _amplitudes_to_validated_state_tensor(state=state_like,
+                                                         qid_shape=qid_shape,
+                                                         dtype=dtype,
+                                                         atol=atol)
+
+        if state_like.shape == (len(qid_shape),):
+            return _qudit_values_to_state_tensor(state=state_like,
+                                                 qid_shape=qid_shape,
+                                                 dtype=dtype)
+
+        raise ValueError(
+            '`state_like` was convertible to a numpy array, but its '
+            'shape was neither the shape of a list of computational basis '
+            'values (`len(qid_shape)`) nor the shape of a list or tensor of '
+            'state vector amplitudes (`qid_shape` or `(product(qid_shape),)`.\n'
+            '\n'
+            f'qid_shape={qid_shape!r}\n'
+            f'np.array(state_like).shape={state_like.shape}\n'
+            f'np.array(state_like)={state_like}\n')
+
+    raise TypeError(
+        f'Unrecognized type of STATE_LIKE. The given `state_like` was '
+        f'not a computational basis value, list of computational basis values, '
+        f'list of amplitudes, or tensor of amplitudes.\n'
+        f'\n'
+        f'type(state_like)={type(state_like)}\n'
+        f'qid_shape={qid_shape!r}')
+
+
+def _amplitudes_to_validated_state_tensor(*, state: np.ndarray,
+                                          qid_shape: Tuple[int, ...],
+                                          dtype: Type[np.number],
+                                          atol: float) -> np.ndarray:
+    result = np.array(state, dtype=dtype).reshape(qid_shape)
+    validate_normalized_state(result,
                               qid_shape=qid_shape,
                               dtype=dtype,
                               atol=atol)
-    return state
+    return result
+
+
+def _qudit_values_to_state_tensor(*, state: np.ndarray,
+                                  qid_shape: Tuple[int, ...],
+                                  dtype: Type[np.number]) -> np.ndarray:
+
+    for i in range(len(qid_shape)):
+        s = state[i]
+        q = qid_shape[i]
+        if not 0 <= s < q:
+            raise ValueError(
+                f'Qudit value {s} at index {i} is out of bounds for '
+                f'qudit dimension {q}.\n'
+                f'\n'
+                f'qid_shape={qid_shape!r}\n'
+                f'state={state!r}\n')
+
+    if state.dtype.kind[0] not in '?bBiu':
```

Comment:
Is there a better way to check this?

Reply:
I'm honestly not sure. Multiply it by 0, add 0.1, and see if it's still 0?

Reply:
```
>>> issubclass(np.int8, np.integer)
True
>>> issubclass(np.float64, np.integer)
False
```
https://stackoverflow.com/questions/934616/how-do-i-find-out-if-a-numpy-array-contains-integers

Reply:
Checking against `numbers.Integral` should verify that the variable obeys integer arithmetic.
```
>>> issubclass(int, numbers.Integral)
True
>>> issubclass(bool, numbers.Integral)
True
>>> issubclass(np.int8, numbers.Integral)
True
>>> issubclass(np.int16, numbers.Integral)
True
>>> issubclass(np.bool, numbers.Integral)
True
>>> issubclass(float, numbers.Integral)
False
>>> issubclass(np.float64, numbers.Integral)
False
```

Reply:
There seems to be something tricky going on. Although `np.bool` is integral, the dtype type of an array initialized to have dtype `np.bool` is not:

```
>>> issubclass(np.bool, numbers.Integral)                                                                                  
True
```

```
>>> issubclass(np.array([1, 2, 3], dtype=np.bool).dtype.type, numbers.Integral)                                            
False
```

```
>>> issubclass(np.array([1, 2, 3], dtype=np.int32).dtype.type, numbers.Integral)                                            
True
```

Reply:
The trickery appears to be that when you ask for an array of `np.bool` you don't get an array of `np.bool`. Instead you get an array of `np.bool_` which inherits from `np.generic` instead of `np.int`. 
```
>>> test_array = np.array([1, 2, 3], dtype=np.bool)
>>> test_array.dtype.type == np.bool
False
>>> second_test.dtype.type
<class 'numpy.bool_'>
>>> test_array.dtype.type.__bases__
(<class 'numpy.generic'>,)
>>> np.bool.__bases__
(<class 'int'>,)
>>> test_array.dtype.type == np.bool_
True
```

Reply:
`np.bool_`s do not behave like numbers with respect to arithmetic since they do not inherit from `np.number`. I don't know of a way for a numpy array of boolean values to have a dtype other than `np.bool_`. Whenever you index such an array, you will get a `np.bool_` object. In order to make a numpy array give you python `bool`s it seems that you need to use the `item` function. 

You will probably need to add a check for `np.bool_` in addition to the `numbers.Integral` check.
`if not (isinstance(state.dtype, np.bool_) or issubclass(state.dtype, numbers.Integral)):`

If the array has at least one element, an equivalent check would be given by the below code since the `item` function would return a python `bool` which is a `numbers.Integral`.
`if not isinstance(state.item(0), numbers.Integral):` 

 As written, if someone passed in an empty array for `state` and an empty tuple for `qid_shape`, the output of this function will be a `Scalar` of type `dtype`. Is this desired? 

```
_qudit_values_to_state_tensor(state=np.empty(shape=(0),dtype=np.int64),
                                                 qid_shape=tuple(), dtype=np.int64)
array(1)
_qudit_values_to_state_tensor(state=np.empty(shape=(0),dtype=np.int64),
                                                 qid_shape=tuple(), dtype=np.bool)
array(True)
```